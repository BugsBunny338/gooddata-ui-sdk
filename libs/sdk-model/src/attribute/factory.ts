// (C) 2019 GoodData Corporation
import identity = require("lodash/identity");
import { IAttribute } from ".";

/**
 * Builder for attributes.
 *
 * Do not instantiate this class directly. Instead use {@link newAttribute}.
 *
 * @public
 */
export class AttributeBuilder implements IAttribute {
    public attribute: IAttribute["attribute"];

    /**
     * @internal
     */
    constructor(displayFormId: string) {
        this.attribute = {
            displayForm: { identifier: displayFormId },
            localIdentifier: `a_${displayFormId}`,
        };
    }

    public alias = (alias: string) => {
        this.attribute.alias = alias;
        return this;
    };

    public localId = (localId: string) => {
        this.attribute.localIdentifier = localId;
        return this;
    };

    public build = () => {
        return { attribute: this.attribute };
    };
}

/**
 * Function that will be called to perform modifications of an attribute before it is fully constructed.
 *
 * @public
 */
export type AttributeModifications = (builder: AttributeBuilder) => AttributeBuilder;

/**
 * Creates a new attribute with the specified display form identifier and optional modifications and localIdentifier.
 * @param displayFormId - identifier of the attribute display form
 * @param modifications - optional modifications (e.g. alias, etc.)
 * @public
 */
export function newAttribute(
    displayFormId: string,
    modifications: AttributeModifications = identity,
): IAttribute {
    const builder = new AttributeBuilder(displayFormId);
    return modifications(builder).build();
}
