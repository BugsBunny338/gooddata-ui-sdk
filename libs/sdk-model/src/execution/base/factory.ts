// (C) 2020 GoodData Corporation
import { Identifier, Uri, IdentifierRef, UriRef, LocalIdRef } from ".";

/**
 * Creates an IdentifierRef from an identifier
 * @param identifier - identifier to use
 * @public
 */
export function idRef(identifier: Identifier): IdentifierRef {
    return { identifier };
}

/**
 * Creates an UriRef from an URI
 * @param uri - URI to use
 * @public
 */
export function uriRef(uri: Uri): UriRef {
    return { uri };
}

/**
 * Creates an LocalIdRef from a local identifier
 * @param localIdentifier - local identifier to use
 * @public
 */
export function localIdRef(localIdentifier: Identifier): LocalIdRef {
    return { localIdentifier };
}
