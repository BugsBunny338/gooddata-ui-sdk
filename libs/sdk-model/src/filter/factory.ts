// (C) 2019 GoodData Corporation
import {
    AttributeElements,
    IAbsoluteDateFilter,
    INegativeAttributeFilter,
    IPositiveAttributeFilter,
    IRelativeDateFilter,
} from ".";
import { attributeIdentifier, IAttribute } from "../attribute";

/**
 * Creates a new positive attribute filter.
 *
 * @param attributeOrId - either instance of attribute to create filter for or identifier of attribute's display form
 *  if the input is attribute object, then it is expected that attribute references display form by identifier
 * @param inValues - values to filter for; these can be either specified as AttributeElements object or as an array
 *  of attribute element _values_
 * @public
 */
export function newPositiveAttributeFilter(
    attributeOrId: IAttribute | string,
    inValues: AttributeElements | string[],
): IPositiveAttributeFilter {
    const displayFormId =
        typeof attributeOrId === "string" ? attributeOrId : attributeIdentifier(attributeOrId)!;
    const inObject: AttributeElements = Array.isArray(inValues) ? { values: inValues } : inValues;

    return {
        positiveAttributeFilter: {
            displayForm: { identifier: displayFormId },
            in: inObject,
        },
    };
}

/**
 * Creates a new negative attribute filter.
 *
 * @param attributeOrId - either instance of attribute to create filter for or identifier of attribute's display form;
 *  if the input is attribute object, then it is expected that attribute references display form by identifier
 * @param notInValues - values to filter out; these can be either specified as AttributeElements object or as an array
 *  of attribute element _values_
 * @public
 */
export function newNegativeAttributeFilter(
    attributeOrId: IAttribute | string,
    notInValues: AttributeElements | string[],
): INegativeAttributeFilter {
    const displayFormId =
        typeof attributeOrId === "string" ? attributeOrId : attributeIdentifier(attributeOrId)!;
    const notInObject: AttributeElements = Array.isArray(notInValues) ? { values: notInValues } : notInValues;

    return {
        negativeAttributeFilter: {
            displayForm: { identifier: displayFormId },
            notIn: notInObject,
        },
    };
}

/**
 * Creates a new absolute date filter.
 *
 * @param dateDataSetId - identifier of the date data set to filter on
 * @param from - start of the interval in ISO-8601 calendar date format
 * @param to - end of the interval in ISO-8601 calendar date format
 * @public
 */
export function newAbsoluteDateFilter(dateDataSetId: string, from: string, to: string): IAbsoluteDateFilter {
    return {
        absoluteDateFilter: {
            dataSet: { identifier: dateDataSetId },
            from,
            to,
        },
    };
}

/**
 * Defines date data set granularities that can be used in relative date filter.
 *
 * @public
 */
export const DateGranularity = {
    date: "GDC.time.date",
    week: "GDC.time.week_us",
    month: "GDC.time.month",
    quarter: "GDC.time.quarter",
    year: "GDC.time.year",
};

/**
 * Creates a new relative date filter.
 *
 * @param dateDataSetId - identifier of the date data set to filter on
 * @param granularity - granularity of the filters (month, year, etc.)
 * @param from - start of the interval – negative numbers mean the past, zero means today, positive numbers mean the future
 * @param to - end of the interval – negative numbers mean the past, zero means today, positive numbers mean the future
 * @public
 */
export function newRelativeDateFilter(
    dateDataSetId: string,
    granularity: string,
    from: number,
    to: number,
): IRelativeDateFilter {
    return {
        relativeDateFilter: {
            dataSet: { identifier: dateDataSetId },
            granularity,
            from,
            to,
        },
    };
}
