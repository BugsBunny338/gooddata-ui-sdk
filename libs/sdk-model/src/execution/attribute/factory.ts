// (C) 2019-2020 GoodData Corporation
import identity = require("lodash/identity");
import cloneDeep = require("lodash/cloneDeep");
import isEmpty = require("lodash/isEmpty");
import { IAttribute, isAttribute } from "./index";
import { ObjRef, objRefToString, Identifier, isObjRef } from "../base";
import { idRef } from "../base/factory";
import SparkMD5 from "spark-md5";

type AttributeBuilderInput = Identifier | ObjRef | IAttribute;

/**
 * Builder for attributes.
 *
 * Do not instantiate this class directly. Instead use {@link newAttribute}.
 *
 * @public
 */
export class AttributeBuilder {
    private attribute: IAttribute["attribute"];
    private customLocalId: boolean = false;

    /**
     * @internal
     */
    constructor(input: AttributeBuilderInput) {
        if (isAttribute(input)) {
            this.attribute = cloneDeep(input.attribute);
            this.customLocalId = false;
        } else {
            const displayForm: ObjRef = isObjRef(input) ? input : idRef(input, "displayForm");

            this.attribute = {
                displayForm,
                localIdentifier: "",
            };
        }
    }

    public alias = (alias: string) => {
        this.attribute.alias = alias;

        return this;
    };

    public localId = (localId: string) => {
        this.attribute.localIdentifier = localId;
        this.customLocalId = true;

        return this;
    };

    public build = () => {
        const localIdentifier = this.getOrGenerateLocalId();

        return {
            attribute: {
                ...this.attribute,
                localIdentifier,
            },
        };
    };

    private getOrGenerateLocalId(): string {
        if (this.customLocalId && !isEmpty(this.attribute.localIdentifier)) {
            return this.attribute.localIdentifier;
        }

        return ["a", this.calculateAliasHash(), objRefToString(this.attribute.displayForm)]
            .filter(part => !isEmpty(part))
            .join("_");
    }

    private calculateAliasHash(): string {
        if (!this.attribute.alias) {
            return "";
        }

        const hasher = new SparkMD5();
        hasher.append(this.attribute.alias);

        return hasher.end().substr(0, 8);
    }
}

/**
 * Function that will be called to perform modifications of an attribute before it is fully constructed.
 *
 * @public
 */
export type AttributeModifications = (builder: AttributeBuilder) => AttributeBuilder;

/**
 * Creates a new attribute with the specified display form ref and optional modifications and localIdentifier.
 * @param displayFormRefOrId - ref or identifier of the attribute display form
 * @param modifications - optional modifications (e.g. alias, etc.)
 * @public
 */
export function newAttribute(
    displayFormRefOrId: ObjRef | Identifier,
    modifications: AttributeModifications = identity,
): IAttribute {
    const builder = new AttributeBuilder(displayFormRefOrId);

    return modifications(builder).build();
}

/**
 * Allows modification of an existing attribute instance.
 *
 * @param attribute - attribute to modify
 * @param modifications - modification function
 * @public
 */
export function modifyAttribute(
    attribute: IAttribute,
    modifications: AttributeModifications = identity,
): IAttribute {
    const builder = new AttributeBuilder(attribute);

    return modifications(builder).build();
}
