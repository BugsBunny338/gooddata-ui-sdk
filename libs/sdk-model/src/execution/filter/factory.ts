// (C) 2019-2020 GoodData Corporation
import invariant from "ts-invariant";
import {
    AttributeElements,
    ComparisonConditionOperator,
    IAbsoluteDateFilter,
    IMeasureValueFilter,
    INegativeAttributeFilter,
    IPositiveAttributeFilter,
    IRelativeDateFilter,
    RangeConditionOperator,
} from "./index";
import { IAttribute, attributeAttributeDisplayFormObjRef } from "../attribute";
import { ObjRefInScope, ObjRef, isObjRef, Identifier } from "../base";
import { IMeasure, isMeasure, measureLocalId } from "../measure";
import { idRef } from "../base/factory";

/**
 * Creates a new positive attribute filter.
 *
 * @param attributeOrRef - either instance of attribute to create filter for or ref or identifier of attribute's display form
 * @param inValues - values to filter for; these can be either specified as AttributeElements object or as an array
 *  of attribute element _values_
 * @public
 */
export function newPositiveAttributeFilter(
    attributeOrRef: IAttribute | ObjRef | Identifier,
    inValues: AttributeElements | string[],
): IPositiveAttributeFilter {
    const objRef = isObjRef(attributeOrRef)
        ? attributeOrRef
        : typeof attributeOrRef === "string"
        ? idRef(attributeOrRef)
        : attributeAttributeDisplayFormObjRef(attributeOrRef);

    const inObject: AttributeElements = Array.isArray(inValues) ? { values: inValues } : inValues;

    return {
        positiveAttributeFilter: {
            displayForm: objRef,
            in: inObject,
        },
    };
}

/**
 * Creates a new negative attribute filter.
 *
 * @param attributeOrRef - either instance of attribute to create filter for or ref or identifier of attribute's display form
 * @param notInValues - values to filter out; these can be either specified as AttributeElements object or as an array
 *  of attribute element _values_
 * @public
 */
export function newNegativeAttributeFilter(
    attributeOrRef: IAttribute | ObjRef | Identifier,
    notInValues: AttributeElements | string[],
): INegativeAttributeFilter {
    const objRef = isObjRef(attributeOrRef)
        ? attributeOrRef
        : typeof attributeOrRef === "string"
        ? idRef(attributeOrRef)
        : attributeAttributeDisplayFormObjRef(attributeOrRef);

    const notInObject: AttributeElements = Array.isArray(notInValues) ? { values: notInValues } : notInValues;

    return {
        negativeAttributeFilter: {
            displayForm: objRef,
            notIn: notInObject,
        },
    };
}

/**
 * Creates a new absolute date filter.
 *
 * @param dateDataSet - ref or identifier of the date data set to filter on
 * @param from - start of the interval in ISO-8601 calendar date format
 * @param to - end of the interval in ISO-8601 calendar date format
 * @public
 */
export function newAbsoluteDateFilter(
    dateDataSet: ObjRef | Identifier,
    from: string,
    to: string,
): IAbsoluteDateFilter {
    const dataSet = isObjRef(dateDataSet) ? dateDataSet : idRef(dateDataSet);
    return {
        absoluteDateFilter: {
            dataSet,
            from,
            to,
        },
    };
}

/**
 * Creates a new relative date filter.
 *
 * @param dateDataSet - ref or identifier of the date data set to filter on
 * @param granularity - granularity of the filters (month, year, etc.)
 * @param from - start of the interval – negative numbers mean the past, zero means today, positive numbers mean the future
 * @param to - end of the interval – negative numbers mean the past, zero means today, positive numbers mean the future
 * @public
 */
export function newRelativeDateFilter(
    dateDataSet: ObjRef | Identifier,
    granularity: string,
    from: number,
    to: number,
): IRelativeDateFilter {
    const dataSet = isObjRef(dateDataSet) ? dateDataSet : idRef(dateDataSet);
    return {
        relativeDateFilter: {
            dataSet,
            granularity,
            from,
            to,
        },
    };
}

/**
 * Creates a new measure value filter.
 *
 * @param measureOrRef - instance of measure to filter, or reference of the measure object; if instance of measure is
 *   provided, then it is assumed this measure is in scope of execution and will be referenced by the filter by
 *   its local identifier
 * @param operator - comparison or range operator to use in the filter
 * @param val1 - first numeric value, used as value in comparison or as 'from' value in range operators
 * @param val2 - second numeric value, required in range operators and used in 'to' value; optional and ignored with comparison operators
 * @public
 */
export function newMeasureValueFilter(
    measureOrRef: IMeasure | ObjRefInScope,
    operator: ComparisonConditionOperator | RangeConditionOperator,
    val1: number,
    val2?: number,
): IMeasureValueFilter {
    const ref: ObjRefInScope = isMeasure(measureOrRef)
        ? { localIdentifier: measureLocalId(measureOrRef) }
        : measureOrRef;

    if (operator === "BETWEEN" || operator === "NOT_BETWEEN") {
        invariant(val2, "measure value filter with range operator requires two numeric values");

        return {
            measureValueFilter: {
                measure: ref,
                condition: {
                    range: {
                        operator,
                        from: val1,
                        to: val2!,
                    },
                },
            },
        };
    } else {
        return {
            measureValueFilter: {
                measure: ref,
                condition: {
                    comparison: {
                        operator,
                        value: val1,
                    },
                },
            },
        };
    }
}
