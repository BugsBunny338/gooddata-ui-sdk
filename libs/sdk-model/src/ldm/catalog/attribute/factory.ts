// (C) 2019-2020 GoodData Corporation
import identity = require("lodash/identity");
import { BuilderModifications, builderFactory } from "../../../base/builder";
import { ICatalogAttribute } from ".";
import { IAttributeMetadataObject } from "../../metadata/attribute";
import { IAttributeDisplayFormMetadataObject } from "../../metadata/attributeDisplayForm";
import { GroupableCatalogItemBuilder } from "../group/factory";
import { newAttributeMetadataObject, AttributeMetadataObjectBuilder } from "../../metadata/attribute/factory";
import { ObjRef, isObjRef } from "../../../objRef";
import {
    AttributeDisplayFormMetadataObjectBuilder,
    newAttributeDisplayFormMetadataObject,
} from "../../metadata";
import { AttributeModifications, newAttribute } from "../../../execution/attribute/factory";
import { IAttribute } from "../../../execution/attribute";

/**
 * @public
 */
export class CatalogAttributeBuilder<
    T extends ICatalogAttribute = ICatalogAttribute
> extends GroupableCatalogItemBuilder<T> {
    public attribute(
        attributeOrRef: IAttributeMetadataObject | ObjRef,
        modifications?: BuilderModifications<AttributeMetadataObjectBuilder>,
    ): this {
        if (isObjRef(attributeOrRef)) {
            this.item.attribute = newAttributeMetadataObject(attributeOrRef, modifications);
        } else {
            this.item.attribute = attributeOrRef;
        }
        return this;
    }

    public defaultDisplayForm(
        displayFormOrRef: IAttributeDisplayFormMetadataObject | ObjRef,
        modifications?: BuilderModifications<AttributeDisplayFormMetadataObjectBuilder>,
    ): this {
        if (isObjRef(displayFormOrRef)) {
            this.item.defaultDisplayForm = newAttributeDisplayFormMetadataObject(
                displayFormOrRef,
                modifications,
            );
        } else {
            this.item.defaultDisplayForm = displayFormOrRef;
        }
        return this;
    }

    public toExecutionModel(modifications: AttributeModifications = identity): IAttribute {
        if (!this.item.defaultDisplayForm) {
            throw new Error("Cannot convert catalog attribute to execution model, no displayForm found!");
        }

        const defaultModifications: AttributeModifications = a =>
            a.alias(this.item.defaultDisplayForm?.title!);

        return newAttribute(this.item.defaultDisplayForm.ref, m => modifications(defaultModifications(m)));
    }
}

/**
 * @public
 */
export const newCatalogAttribute = (
    modifications: BuilderModifications<CatalogAttributeBuilder> = identity,
): ICatalogAttribute => builderFactory(CatalogAttributeBuilder, { type: "attribute" }, modifications);

/**
 * @public
 */
export const catalogAttributeToExecutionAttribute = (
    catalogAttribute: ICatalogAttribute,
    modifications: BuilderModifications<CatalogAttributeBuilder> = identity,
): IAttribute => new CatalogAttributeBuilder(catalogAttribute).modify(modifications).toExecutionModel();
