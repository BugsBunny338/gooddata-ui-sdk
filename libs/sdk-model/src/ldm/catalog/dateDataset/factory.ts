// (C) 2019-2020 GoodData Corporation
import identity = require("lodash/identity");
import { BuilderModifications, builderFactory, Builder } from "../../../base/builder";
import { ICatalogDateDataset, ICatalogDateAttribute, CatalogDateAttributeGranularity } from ".";
import { IAttributeMetadataObject } from "../../metadata/attribute";
import { IAttributeDisplayFormMetadataObject } from "../../metadata/attributeDisplayForm";
import { isObjRef, ObjRef } from "../../../objRef";
import { newAttributeMetadataObject, AttributeMetadataObjectBuilder } from "../../metadata/attribute/factory";
import {
    AttributeDisplayFormMetadataObjectBuilder,
    newAttributeDisplayFormMetadataObject,
} from "../../metadata/attributeDisplayForm/factory";
import {
    DataSetMetadataObjectBuilder,
    newDataSetMetadataObject,
    IDataSetMetadataObject,
} from "../../metadata";

/**
 * Catalog date attribute builder
 * See {@link Builder}
 *
 * @public
 */
export class CatalogDateAttributeBuilder<
    T extends ICatalogDateAttribute = ICatalogDateAttribute
> extends Builder<T> {
    public granularity(granularity: CatalogDateAttributeGranularity): this {
        this.item.granularity = granularity;
        return this;
    }

    public attribute(
        attributeOrRef: IAttributeMetadataObject | ObjRef,
        modifications?: BuilderModifications<AttributeMetadataObjectBuilder>,
    ): this {
        if (isObjRef(attributeOrRef)) {
            this.item.attribute = newAttributeMetadataObject(attributeOrRef, modifications);
        } else {
            this.item.attribute = attributeOrRef;
        }
        return this;
    }

    public defaultDisplayForm(
        displayFormOrRef: IAttributeDisplayFormMetadataObject | ObjRef,
        modifications?: BuilderModifications<AttributeDisplayFormMetadataObjectBuilder>,
    ): this {
        if (isObjRef(displayFormOrRef)) {
            this.item.defaultDisplayForm = newAttributeDisplayFormMetadataObject(
                displayFormOrRef,
                modifications,
            );
        } else {
            this.item.defaultDisplayForm = displayFormOrRef;
        }
        return this;
    }
}

/**
 * Catalog date attribute factory
 *
 * @param modifications - catalog date attribute builder modifications to perform
 * @returns created catalog date attribute
 * @public
 */
export const newCatalogDateAttribute = (
    modifications: BuilderModifications<CatalogDateAttributeBuilder> = identity,
): ICatalogDateAttribute => builderFactory(CatalogDateAttributeBuilder, {}, modifications);

/**
 * Catalog date dataset builder
 * See {@link Builder}
 *
 * @public
 */
export class CatalogDateDatasetBuilder<T extends ICatalogDateDataset = ICatalogDateDataset> extends Builder<
    T
> {
    public relevance(relevance: number): this {
        this.item.relevance = relevance;
        return this;
    }

    public dateAttributes(dateAttributes: ICatalogDateAttribute[]): this {
        this.item.dateAttributes = dateAttributes;
        return this;
    }

    public dataSet(
        dataSetOrRef: IDataSetMetadataObject | ObjRef,
        modifications?: BuilderModifications<DataSetMetadataObjectBuilder>,
    ): this {
        if (isObjRef(dataSetOrRef)) {
            this.item.dataSet = newDataSetMetadataObject(dataSetOrRef, modifications);
        } else {
            this.item.dataSet = dataSetOrRef;
        }
        return this;
    }
}

/**
 * Catalog date dataset factory
 *
 * @param modifications - catalog date dataset builder modifications to perform
 * @returns created catalog date dataset
 * @public
 */
export const newCatalogDateDataset = (
    modifications: BuilderModifications<CatalogDateDatasetBuilder> = identity,
): ICatalogDateDataset => builderFactory(CatalogDateDatasetBuilder, { type: "dateDataset" }, modifications);
