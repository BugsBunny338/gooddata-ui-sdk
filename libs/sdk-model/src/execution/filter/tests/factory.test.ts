// (C) 2019-2020 GoodData Corporation
import { Account, Won } from "../../../../__mocks__/model";
import {
    newAbsoluteDateFilter,
    newMeasureValueFilter,
    newNegativeAttributeFilter,
    newPositiveAttributeFilter,
    newRelativeDateFilter,
} from "../factory";
import { localIdRef } from "../../../objRef/factory";

describe("filter factory", () => {
    describe("newPositiveAttributeFilter", () => {
        it("should generate correct uri filter", () => {
            expect(newPositiveAttributeFilter("foo", { uris: ["bar", "baz"] })).toMatchSnapshot();
        });

        it("should generate correct value filter", () => {
            expect(newPositiveAttributeFilter("foo", { values: ["bar", "baz"] })).toMatchSnapshot();
        });

        it("should generate correct value filter when input is array of values", () => {
            expect(newPositiveAttributeFilter("foo", ["bar", "baz"])).toMatchSnapshot();
        });

        it("should generate correct filter attribute object is on input", () => {
            expect(newPositiveAttributeFilter(Account.Name, ["bar", "baz"])).toMatchSnapshot();
        });
    });

    describe("newNegativeAttributeFilter", () => {
        it("should generate correct filter", () => {
            expect(newNegativeAttributeFilter("foo", { uris: ["bar", "baz"] })).toMatchSnapshot();
        });
        it("should generate correct textual filter", () => {
            expect(newNegativeAttributeFilter("foo", { values: ["bar", "baz"] })).toMatchSnapshot();
        });
        it("should generate correct textual filter when input is array of values", () => {
            expect(newNegativeAttributeFilter("foo", ["bar", "baz"])).toMatchSnapshot();
        });
        it("should generate correct filter when input is IAttribute instance", () => {
            expect(newNegativeAttributeFilter(Account.Name, ["bar", "baz"])).toMatchSnapshot();
        });
    });

    describe("newAbsoluteDateFilter", () => {
        it("should generate correct filter", () => {
            expect(newAbsoluteDateFilter("foo", "2018-01-01", "2018-12-31")).toMatchSnapshot();
        });
    });

    describe("newRelativeDateFilter", () => {
        it("should generate correct filter", () => {
            expect(newRelativeDateFilter("foo", "GDC.time.month", 1, 3)).toMatchSnapshot();
        });
    });

    describe("newMeasureValueFilter", () => {
        it("should generate comparison filter for measure object", () => {
            expect(newMeasureValueFilter(Won, "EQUAL_TO", 11)).toMatchSnapshot();
        });
        it("should generate comparison filter for measure identifier", () => {
            expect(newMeasureValueFilter(localIdRef("measureObjLocalId"), "EQUAL_TO", 11)).toMatchSnapshot();
        });
        it("should generate comparison filter for measure local identifier", () => {
            expect(newMeasureValueFilter(localIdRef("measureObjLocalId"), "EQUAL_TO", 11)).toMatchSnapshot();
        });
        it("should generate comparison filter for measure object with treatNullValuesAs", () => {
            expect(newMeasureValueFilter(Won, "EQUAL_TO", 11, 42)).toMatchSnapshot();
        });
        it("should generate range filter for measure object", () => {
            expect(newMeasureValueFilter(Won, "BETWEEN", 0, 100)).toMatchSnapshot();
        });
        it("should generate ranger filter for measure identifier", () => {
            expect(
                newMeasureValueFilter(localIdRef("measureObjLocalId"), "BETWEEN", 0, 100),
            ).toMatchSnapshot();
        });
        it("should generate ranger filter for measure local identifier", () => {
            expect(newMeasureValueFilter("measureObjLocalId", "BETWEEN", 0, 100)).toMatchSnapshot();
        });
        it("should generate range filter for measure object with treatNullValuesAs", () => {
            expect(newMeasureValueFilter(Won, "BETWEEN", 0, 100, 42)).toMatchSnapshot();
        });
    });
});
