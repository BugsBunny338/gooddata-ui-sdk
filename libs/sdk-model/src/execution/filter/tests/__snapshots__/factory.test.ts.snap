// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter factory newAbsoluteDateFilter should generate correct filter 1`] = `
Object {
  "absoluteDateFilter": Object {
    "dataSet": Object {
      "identifier": "foo",
    },
    "from": "2018-01-01",
    "to": "2018-12-31",
  },
}
`;

exports[`filter factory newMeasureValueFilter should generate comparison filter for measure identifier 1`] = `
Object {
  "measureValueFilter": Object {
    "condition": Object {
      "comparison": Object {
        "operator": "EQUAL_TO",
        "value": 11,
      },
    },
    "measure": Object {
      "localIdentifier": "measureObjLocalId",
    },
  },
}
`;

exports[`filter factory newMeasureValueFilter should generate comparison filter for measure local identifier 1`] = `
Object {
  "measureValueFilter": Object {
    "condition": Object {
      "comparison": Object {
        "operator": "EQUAL_TO",
        "value": 11,
      },
    },
    "measure": Object {
      "localIdentifier": "measureObjLocalId",
    },
  },
}
`;

exports[`filter factory newMeasureValueFilter should generate comparison filter for measure object 1`] = `
Object {
  "measureValueFilter": Object {
    "condition": Object {
      "comparison": Object {
        "operator": "EQUAL_TO",
        "value": 11,
      },
    },
    "measure": Object {
      "localIdentifier": "m_afSEwRwdbMeQ",
    },
  },
}
`;

exports[`filter factory newMeasureValueFilter should generate comparison filter for measure object with treatNullValuesAs 1`] = `
Object {
  "measureValueFilter": Object {
    "condition": Object {
      "comparison": Object {
        "operator": "EQUAL_TO",
        "treatNullValuesAs": 42,
        "value": 11,
      },
    },
    "measure": Object {
      "localIdentifier": "m_afSEwRwdbMeQ",
    },
  },
}
`;

exports[`filter factory newMeasureValueFilter should generate range filter for measure object 1`] = `
Object {
  "measureValueFilter": Object {
    "condition": Object {
      "range": Object {
        "from": 0,
        "operator": "BETWEEN",
        "to": 100,
      },
    },
    "measure": Object {
      "localIdentifier": "m_afSEwRwdbMeQ",
    },
  },
}
`;

exports[`filter factory newMeasureValueFilter should generate range filter for measure object with treatNullValuesAs 1`] = `
Object {
  "measureValueFilter": Object {
    "condition": Object {
      "range": Object {
        "from": 0,
        "operator": "BETWEEN",
        "to": 100,
        "treatNullValuesAs": 42,
      },
    },
    "measure": Object {
      "localIdentifier": "m_afSEwRwdbMeQ",
    },
  },
}
`;

exports[`filter factory newMeasureValueFilter should generate ranger filter for measure identifier 1`] = `
Object {
  "measureValueFilter": Object {
    "condition": Object {
      "range": Object {
        "from": 0,
        "operator": "BETWEEN",
        "to": 100,
      },
    },
    "measure": Object {
      "localIdentifier": "measureObjLocalId",
    },
  },
}
`;

exports[`filter factory newMeasureValueFilter should generate ranger filter for measure local identifier 1`] = `
Object {
  "measureValueFilter": Object {
    "condition": Object {
      "range": Object {
        "from": 0,
        "operator": "BETWEEN",
        "to": 100,
      },
    },
    "measure": Object {
      "localIdentifier": "measureObjLocalId",
    },
  },
}
`;

exports[`filter factory newNegativeAttributeFilter should generate correct filter 1`] = `
Object {
  "negativeAttributeFilter": Object {
    "displayForm": Object {
      "identifier": "foo",
    },
    "notIn": Object {
      "uris": Array [
        "bar",
        "baz",
      ],
    },
  },
}
`;

exports[`filter factory newNegativeAttributeFilter should generate correct filter when input is IAttribute instance 1`] = `
Object {
  "negativeAttributeFilter": Object {
    "displayForm": Object {
      "identifier": "label.account.id.name",
      "type": "displayForm",
    },
    "notIn": Object {
      "values": Array [
        "bar",
        "baz",
      ],
    },
  },
}
`;

exports[`filter factory newNegativeAttributeFilter should generate correct textual filter 1`] = `
Object {
  "negativeAttributeFilter": Object {
    "displayForm": Object {
      "identifier": "foo",
    },
    "notIn": Object {
      "values": Array [
        "bar",
        "baz",
      ],
    },
  },
}
`;

exports[`filter factory newNegativeAttributeFilter should generate correct textual filter when input is array of values 1`] = `
Object {
  "negativeAttributeFilter": Object {
    "displayForm": Object {
      "identifier": "foo",
    },
    "notIn": Object {
      "values": Array [
        "bar",
        "baz",
      ],
    },
  },
}
`;

exports[`filter factory newPositiveAttributeFilter should generate correct filter attribute object is on input 1`] = `
Object {
  "positiveAttributeFilter": Object {
    "displayForm": Object {
      "identifier": "label.account.id.name",
      "type": "displayForm",
    },
    "in": Object {
      "values": Array [
        "bar",
        "baz",
      ],
    },
  },
}
`;

exports[`filter factory newPositiveAttributeFilter should generate correct uri filter 1`] = `
Object {
  "positiveAttributeFilter": Object {
    "displayForm": Object {
      "identifier": "foo",
    },
    "in": Object {
      "uris": Array [
        "bar",
        "baz",
      ],
    },
  },
}
`;

exports[`filter factory newPositiveAttributeFilter should generate correct value filter 1`] = `
Object {
  "positiveAttributeFilter": Object {
    "displayForm": Object {
      "identifier": "foo",
    },
    "in": Object {
      "values": Array [
        "bar",
        "baz",
      ],
    },
  },
}
`;

exports[`filter factory newPositiveAttributeFilter should generate correct value filter when input is array of values 1`] = `
Object {
  "positiveAttributeFilter": Object {
    "displayForm": Object {
      "identifier": "foo",
    },
    "in": Object {
      "values": Array [
        "bar",
        "baz",
      ],
    },
  },
}
`;

exports[`filter factory newRelativeDateFilter should generate correct filter 1`] = `
Object {
  "relativeDateFilter": Object {
    "dataSet": Object {
      "identifier": "foo",
    },
    "from": 1,
    "granularity": "GDC.time.month",
    "to": 3,
  },
}
`;
